src/_ttconv.cpp

```diff
@@ -5,157 +5,79 @@
+static void convert_ttf_to_ps(
+    const char *filename,
+    pybind11::object &output,
+    int fonttype,
+    pybind11::iterable* glyph_ids)
 {
-    PythonFileWriter *file_writer = (PythonFileWriter *)address;
-
-    PyObject *write_method = PyObject_GetAttrString(object, "write");
-    if (write_method == NULL || !PyCallable_Check(write_method)) {
-        PyErr_SetString(PyExc_TypeError, "Expected a file-like object with a write method.");
-        return 0;
-    }
+    PythonFileWriter output_(output);
 
-    file_writer->set(write_method);
-    Py_DECREF(write_method);
-
-    return 1;
-}
-
-int pyiterable_to_vector_int(PyObject *object, void *address)
-{
-    std::vector<int> *result = (std::vector<int> *)address;
-
-    PyObject *iterator = PyObject_GetIter(object);
-    if (!iterator) {
-        return 0;
-    }
-
-    PyObject *item;
-    while ((item = PyIter_Next(iterator))) {
-        long value = PyLong_AsLong(item);
-        Py_DECREF(item);
-        if (value == -1 && PyErr_Occurred()) {
-            return 0;
+    std::vector<int> glyph_ids_;
+    if (glyph_ids) {
+        for (pybind11::handle glyph_id: *glyph_ids) {
+            glyph_ids_.push_back(glyph_id.cast<int>());
         }
-        result->push_back((int)value);
     }
 
-    Py_DECREF(iterator);
-
-    return 1;
-}
-
-static PyObject *convert_ttf_to_ps(PyObject *self, PyObject *args, PyObject *kwds)
-{
-    const char *filename;
-    PythonFileWriter output;
-    int fonttype;
-    std::vector<int> glyph_ids;
-
-    static const char *kwlist[] = { "filename", "output", "fonttype", "glyph_ids", NULL };
-    if (!PyArg_ParseTupleAndKeywords(args,
-                                     kwds,
-                                     "yO&i|O&:convert_ttf_to_ps",
-                                     (char **)kwlist,
-                                     &filename,
-                                     fileobject_to_PythonFileWriter,
-                                     &output,
-                                     &fonttype,
-                                     pyiterable_to_vector_int,
-                                     &glyph_ids)) {
-        return NULL;
-    }
 
     if (fonttype != 3 && fonttype != 42) {
-        PyErr_SetString(PyExc_ValueError,
-                        "fonttype must be either 3 (raw Postscript) or 42 "
-                        "(embedded Truetype)");
-        return NULL;
+        throw pybind11::value_error(
+            "fonttype must be either 3 (raw Postscript) or 42 (embedded Truetype)");
     }
 
     try
     {
-        insert_ttfont(filename, output, (font_type_enum)fonttype, glyph_ids);
+        insert_ttfont(filename, output_, static_cast<font_type_enum>(fonttype), glyph_ids_);
     }
     catch (TTException &e)
     {
-        PyErr_SetString(PyExc_RuntimeError, e.getMessage());
-        return NULL;
-    }
-    catch (const py::exception &)
-    {
-        return NULL;
+        throw std::runtime_error(e.getMessage());
```

Comment:
I am concerned that we don't exercise this line in the test suite as it would be good to confirm that `TTException`s propagate through `pybind11` as expected. But given that we've evidently never had such a test it doesn't have to be done here, it can be punted to "future work".